src/test/java/org/apache/commons/validator/routines/checkdigit/AbstractCheckDigitTest.java

```diff
@@ -154,7 +154,11 @@ public void testCalculateInvalid() {
                     log.debug("   " + i + " Testing Invalid Check Digit, Code=[" + code + "]");
                 }
                 final String expected = checkDigit(code);
-                final String actual = routine.calculate(removeCheckDigit(code));
+                String codeWithNoCheckDigit = removeCheckDigit(code);
+                if (codeWithNoCheckDigit == null) {
+                    throw new CheckDigitException("Invalid Code=[" + code + "]");
```

Comment:
Why not just call `fail()` here?

Reply:
the default invalid test in `class AbstractCheckDigitTest` is
```
protected String[] invalid = { "12345678A" };
```

method `removeCheckDigit` returns null if `REGEX_VALIDATOR.validate(code)` == null. This is the case for "12345678A" because this code is invalid due to its regex.

When using `fail()` as you suggests the test `testCalculateInvalid()` would fail 
but it should `assertTrue(e.getMessage().startsWith("Invalid ") ...` because "12345678A"  IS INVALID!

So we `throw new CheckDigitException("Invalid Code=[" + code + "]");` and catch it in
```
            } catch (final CheckDigitException e) {
                // possible failure messages:
                // Invalid ISBN Length ...
                // Invalid Character[ ...
                // Are there any others?
                assertTrue(e.getMessage().startsWith("Invalid "), "Invalid Character[" + i + "]=" + e.getMessage());
```

NB: `testCalculateInvalid()` is not my idea. I needed a while to understand how it works :-)

Regards